### PR TITLE
board: arm: added support for nucleo_f410rb

### PR DESCRIPTION
**Pull request by hansu**
I added support for the nucleo_f410rb board.
Only one little ugly workaround is included in [clock_stm32_ll_common.c](https://github.com/zephyrproject-rtos/zephyr/compare/master...Hans470:add-nucleo_f410rb#diff-97454ddaf02527eda9076a96ddbb232770e581cade68e324e6c48ce4e033ae04) because this device has no AHB2 bus:
``` diff
	case STM32_CLOCK_BUS_AHB2:
+		#ifdef RCC_AHB2_SUPPORT
		LL_AHB2_GRP1_DisableClock(pclken->enr);
+		#endif
```
May someone have a better idea.

### CODE REVIEW

File: boards/arm/nucleo_f411re/nucleo_f411re.yaml [hunk 1]
```diff
@@ -15,5 +15,5 @@ supported:
   - gpio
   - spi
   - i2c
-ram: 96
+ram: 128
```
Comment (erwango) [inline boards/arm/nucleo_f411re/nucleo_f411re.yaml:18]: Please move this in a dedicated PR

File: doc/conf.py [hunk 2]
```diff
@@ -372,6 +372,7 @@
     ('boards/arm/nucleo_f207zg/doc/nucleo_f207zg', 'boards/arm/nucleo_f207zg/doc/index'),
     ('boards/arm/nucleo_f334r8/doc/nucleof334r8', 'boards/arm/nucleo_f334r8/doc/index'),
     ('boards/arm/nucleo_f401re/doc/nucleof401re', 'boards/arm/nucleo_f401re/doc/index'),
+    ('boards/arm/nucleo_f410rb/doc/nucleof410rb', 'boards/arm/nucleo_f410rb/doc/index'),
```
Comment (erwango) [inline doc/conf.py:375]: This change is not required for newly introduced boards

File: boards/arm/nucleo_f446re/nucleo_f446re.yaml [hunk 3]
```diff
@@ -14,5 +14,5 @@ supported:
   - gpio
   - spi
   - i2c
-ram: 96
+ram: 128
```
Comment (erwango) [inline boards/arm/nucleo_f446re/nucleo_f446re.yaml:17]: In that case, SoC sram is split in 2 areas (SRAM1 112K and SRAM2 16K), so this is correct.

File: drivers/interrupt_controller/intc_exti_stm32.c [hunk 4]
```diff
@@ -560,10 +569,12 @@ static void __stm32_exti_connect_irqs(const struct device *dev)
 		CONFIG_EXTI_STM32_PVD_IRQ_PRI,
 		__stm32_exti_isr_16, DEVICE_GET(exti_stm32),
 		0);
+	#if !defined(CONFIG_SOC_STM32F410RX)
```
Comment (erwango) [inline drivers/interrupt_controller/intc_exti_stm32.c:572]: Please don't indent #ifdef's

File: drivers/interrupt_controller/intc_exti_stm32.c [hunk 5]
```diff
@@ -53,6 +53,15 @@ const IRQn_Type exti_irq_table[] = {
...
 };
+#elif defined(CONFIG_SOC_STM32F410RX) /* STM32F410RX has no OTG_FS_WKUP_IRQn */
+const IRQn_Type exti_irq_table[] = {
+	EXTI0_IRQn, EXTI1_IRQn, EXTI2_IRQn, EXTI3_IRQn,
+	EXTI4_IRQn, EXTI9_5_IRQn, EXTI9_5_IRQn, EXTI9_5_IRQn,
+	EXTI9_5_IRQn, EXTI9_5_IRQn, EXTI15_10_IRQn, EXTI15_10_IRQn,
+	EXTI15_10_IRQn, EXTI15_10_IRQn, EXTI15_10_IRQn, EXTI15_10_IRQn,
+	PVD_IRQn, 0xFF, 0xFF, 0xFF,
+	0xFF, TAMP_STAMP_IRQn, RTC_WKUP_IRQn
+};
```
Comment (erwango) [inline drivers/interrupt_controller/intc_exti_stm32.c:64]: This change should be moved to a dedicated driver commit

File: soc/arm/st_stm32/stm32f4/Kconfig.defconfig.stm32f410xx [hunk 6]
```diff
@@ -0,0 +1,14 @@
+# ST STM32F410XE MCU configuration options
```
Comment (erwango) [inline soc/arm/st_stm32/stm32f4/Kconfig.defconfig.stm32f410xx:1]: STM32F410XB

File: dts/arm/st/f4/stm32f410Xb.dtsi [hunk 7]
```diff
@@ -0,0 +1,23 @@
+/*
+ * Copyright (c) 2018 Linaro Limited
```
Comment (erwango) [inline dts/arm/st/f4/stm32f410Xb.dtsi:2]: Please remove this

File: dts/arm/st/f4/stm32f410.dtsi [hunk 8]
```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2018 Philémon Jaermann
```
Comment (erwango) [inline dts/arm/st/f4/stm32f410.dtsi:2]: Please remove

File: drivers/clock_control/clock_stm32_ll_common.c [hunk 9]
```diff
@@ -101,7 +101,9 @@ static inline int stm32_clock_control_on(const struct device *dev,
 	defined(CONFIG_SOC_SERIES_STM32WBX) || \
 	defined(CONFIG_SOC_SERIES_STM32G4X)
 	case STM32_CLOCK_BUS_AHB2:
+	#ifdef RCC_AHB2_SUPPORT
```
Comment (erwango) [inline drivers/clock_control/clock_stm32_ll_common.c:104]: Please move to a dedicated commit.
Then, `RCC_AHB2_SUPPORT` is defined only on F4, so this works in your case, but will break other cases.
I propose the folllowing:
```
#if defined(CONFIG_SOC_SERIES_STM32L4X) || \
	defined(CONFIG_SOC_SERIES_STM32L5X) || \
	defined(CONFIG_SOC_SERIES_STM32F4X) && !defined(CONFIG_SOC_STM32F410RX) || \
	defined(CONFIG_SOC_SERIES_STM32F7X) || \
	defined(CONFIG_SOC_SERIES_STM32F2X) || \
	defined(CONFIG_SOC_SERIES_STM32WBX) || \
	defined(CONFIG_SOC_SERIES_STM32G4X)
        case STM32_CLOCK_BUS_AHB2:
        [...]
```

**Review by erwango — COMMENTED** [review]
Thanks for this contribution.

Please look at requested changes.
Also, please add sign-off-by in commit messages

File: drivers/interrupt_controller/intc_exti_stm32.c (hunk 5 above)
Comment (Hans470) [reply drivers/interrupt_controller/intc_exti_stm32.c:64]: Thanks for rewiewing. How should I manage that? Do a new PR?

File: boards/arm/nucleo_f446re/nucleo_f446re.yaml (hunk 3 above)
Comment (Hans470) [reply boards/arm/nucleo_f446re/nucleo_f446re.yaml:17]: But in [stm32f446Xe.dtsi](https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/arm/st/f4/stm32f446Xe.dtsi#L12) it is defined as 128k...

File: drivers/interrupt_controller/intc_exti_stm32.c (hunk 5 above)
Comment (erwango) [reply drivers/interrupt_controller/intc_exti_stm32.c:64]: No need, you can split current commit and force push current branch.

Comment (Hans470) [reply drivers/interrupt_controller/intc_exti_stm32.c:64]: But modifying commits and force push as method of choice? Ok I can do...

File: boards/arm/nucleo_f410rb/Kconfig.defconfig [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+if WATCHDOG
+
+config WWDG_STM32
+	default y
+
+config IWDG_STM32
+	default n
+
+endif # WATCHDOG
```
[the remark this answers is not in the document]
Comment (Hans470) [reply boards/arm/nucleo_f410rb/Kconfig.defconfig:23]: No, I just had the `nucleo_f401re/Kconfig.defconfig` as template.

[the remark this answers is not in the document]
Comment (erwango) [reply boards/arm/nucleo_f410rb/Kconfig.defconfig:23]: ok, let's remove then

File: doc/conf.py (hunk 2 above)
Comment (Hans470) [reply doc/conf.py:375]: Should I remove this?

**Review by erwango — CHANGES_REQUESTED** [review]
Few last comments.

Otherwise, please split soc/dts and board into 2 new commits.

File: drivers/clock_control/clock_stm32_ll_common.c [hunk 11]
```diff
@@ -153,7 +153,9 @@ static inline int stm32_clock_control_off(const struct device *dev,
 	defined(CONFIG_SOC_SERIES_STM32WBX) || \
 	defined(CONFIG_SOC_SERIES_STM32G4X)
 	case STM32_CLOCK_BUS_AHB2:
+		#ifdef RCC_AHB2_SUPPORT
```
[the remark this answers is not in the document]
Comment (Hans470) [reply drivers/clock_control/clock_stm32_ll_common.c]: Sorry I missed that this was on two locations...

File: dts/arm/st/f4/stm32f410.dtsi [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x40003800 0x400>;
+			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00004000>;
+			interrupts = <36 5>;
+			status = "disabled";
+			label = "SPI_2";
+		};
+
+		spi3: spi@40003c00 {
```
Comment (FRASTM) [inline dts/arm/st/f4/stm32f410.dtsi:22]: Can you confirm **SPI3**, as the 0x4000 3c00 is "reserved" in the RM 0401
There ar 3 SPI I/F in the stm32f410xB : SPI1, SPI2, SPI5

File: dts/arm/st/f4/stm32f410.dtsi [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+			interrupts = <36 5>;
+			dmas = <&dma1 4 0 0x400 0x3
+				&dma1 3 0 0x400 0x3>;
+			dma-names = "tx", "rx";
+			status = "disabled";
+			label = "I2S_2";
+		};
+
+		i2s3: i2s@40003c00 {
+			compatible = "st,stm32-i2s";
```
Comment (FRASTM) [inline dts/arm/st/f4/stm32f410.dtsi:58]: see rmk for SPI3

File: boards/arm/nucleo_f410rb/doc/index.rst [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
++-----------+------------+-------------------------------------+
+| UART      | on-chip    | serial port                         |
++-----------+------------+-------------------------------------+
+| PINMUX    | on-chip    | pinmux                              |
++-----------+------------+-------------------------------------+
+| GPIO      | on-chip    | gpio                                |
++-----------+------------+-------------------------------------+
+| PWM       | on-chip    | pwm                                 |
++-----------+------------+-------------------------------------+
+| I2C       | on-chip    | i2c                                 |
```
Comment (FRASTM) [inline boards/arm/nucleo_f410rb/doc/index.rst:84]: | I2S       | on-chip    | i2s                                 |
+-----------+------------+-------------------------------------+
| SPI       | on-chip    | spi                                 |
+-----------+------------+-------------------------------------+

File: dts/arm/st/f4/stm32f410.dtsi (hunk 12 above)
Comment (Hans470) [reply dts/arm/st/f4/stm32f410.dtsi:22]: In the [datasheet](https://www.st.com/resource/en/datasheet/stm32f410rb.pdf) there is a SPI3/I2S3
grafik  

But this is one year older than the RM0401, so maybe it's outdated?

Comment (FRASTM) [reply dts/arm/st/f4/stm32f410.dtsi:22]: Well, I cannot tell you better than ... try

Comment (Hans470) [reply dts/arm/st/f4/stm32f410.dtsi:22]: I think SPI3 is wrong here. Everywhere else is only SPI1, SPI2, SPI5 mentioned.

**Review by erwango — APPROVED** [review]

**Review by FRASTM — APPROVED** [review]
